README.md
# Ryzen SMU

<img src="pics/preview1.jpg" width="400" height="650"> <img src="pics/preview2.jpg" width="400" height="650">

<small>*N.B. Images above shows the potential capabilities of this driver, retrieved using the
[monitor_cpu](userspace/monitor_cpu.c)* utility. Other processors will require tweaking on the
user's part or additional tools.</small>

**Ryzen SMU** is a Linux kernel driver that exposes access to the SMU (System Management Unit) for
certain AMD Ryzen Processors. ***Use at your own risk, the author takes no responsibility over
potential damages that may result from the misuse of this driver or programs dependent upon it.***

The following processor code names are supported:

- [Colfax](https://en.wikichip.org/wiki/amd/cores/colfax)
- [Renoir](https://en.wikichip.org/wiki/amd/cores/renoir)
- [Picasso](https://en.wikichip.org/wiki/amd/cores/picasso)
- [Matisse](https://en.wikichip.org/wiki/amd/cores/matisse)
- [Castle Peak](https://en.wikichip.org/wiki/amd/cores/castle_peak)
- [Raven Ridge](https://en.wikichip.org/wiki/amd/cores/raven_ridge)
- [Raven Ridge 2](https://www.techpowerup.com/gpu-specs/amd-raven-2.g888)
- [Summit Ridge](https://en.wikichip.org/wiki/amd/cores/summit_ridge)
- [Pinnacle Ridge](https://en.wikichip.org/wiki/amd/cores/pinnacle_ridge)
- [Rembrandt](https://en.wikichip.org/wiki/amd/cores/rembrandt)
- [Vermeer](https://en.wikichip.org/wiki/amd/cores/vermeer)
- [Vangogh](https://en.wikichip.org/wiki/amd/cores/vangogh)
- [Cezanne](https://en.wikichip.org/wiki/amd/cores/cezanne)
- [Milan](https://en.wikichip.org/wiki/amd/cores/milan)
- [Dali](https://en.wikichip.org/wiki/amd/cores/dali)

In addition, for the following models, the power metrics/monitoring (PM) table (seen in screenshot)
can also be accessed:

- Vermeer ( Ryzen 5000 Desktop Series )
- Matisse ( Ryzen 3000 Desktop Series )
- Cezanne ( Ryzen 5000[GE] APU Series )
- Renoir ( Ryzen 4000[UGHS] APU Series )
- Picasso ( Ryzen 3000[UGHS] APU Series )
- Raven Ridge ( Ryzen 2000[GEU] APU Series )
- Raven Ridge 2

When loaded, the driver exposes several files under sysfs which can only be read with root
permissions (for obvious reasons) at the root path `/sys/kernel/ryzen_smu_drv`:

- `drv_version`
- `version`
- `mp1_if_version`
- `codename`
- `smu_args`
- `mp1_smu_cmd`
- `smn`
- `rsmu_cmd` (Not present on `Rembrandt`, `Vangogh`)

For supported PM table models where RSMU is also supported, the following files are additionally
exposed:

- `pm_table_version`
- `pm_table_size`
- `pm_table`



## Installation

The kernel module may be installed either by DKMS or manually building and inserting the module.

Be sure kernel headers, dev utilities such as `gcc` and `make` are installed beforehand.

### Ubuntu / Debian

```sh
sudo apt install dkms git build-essential linux-headers-$(uname -r)
git clone https://gitlab.com/leogx9r/ryzen_smu.git
cd ryzen_smu
dpkg-buildpackage -us -uc
sudo dpkg -i ../ryzen-smu_*.deb ../ryzen-smu-dkms_*.deb
```

### Arch Linux

Available on the [AUR](https://aur.archlinux.org/packages/ryzen_smu-dkms-git/).

Install it using your AUR helper of choice, example:

```sh
yay -S ryzen_smu-dkms-git
```

### Stand-alone Installation

The module may be built and inserted into the running kernel manually as follows:

```sh
git clone https://gitlab.com/leogx9r/ryzen_smu.git
cd ryzen_smu

make
sudo insmod ryzen_smu.ko
```

## Auto-Loading On System Startup

Assuming you are using `systemd`, the kernel module may be loaded automatically upon system startup
using [`modules-load.d`](https://www.freedesktop.org/software/systemd/man/modules-load.d.html) by
adding the module name to `/etc/modules-load.d/<name>.conf`

Example:

```sh
# /etc/modules-load.d/ryzen_smu.conf
# Load ryzen_smu driver upon startup
ryzen_smu
```

## Confirming Module Works

Upon loading the module, you should see output in your `dmesg` window listing the SMU version:

```sh
# dmesg

...
[1091.154018] ryzen_smu: CPUID: family 0x17, model 0x71, stepping 0x0, package 0x2
[1091.154385] ryzen_smu: SMU v46.54.0
...
```

After which you can verify the existence of the sysfs files and attempt to read them:

```
# ls -lah /sys/kernel/ryzen_smu_drv
total 0
drwxr-xr-x  2 root root    0 May  7 03:01 ./
drwxr-xr-x 14 root root    0 May  7 03:01 ../
-r--------  1 root root 4.0K May  7 03:12 codename
-r--------  1 root root 4.0K May  7 03:12 drv_version
-r--------  1 root root 4.0K May  7 03:10 mp1_if_version
-rw-------  1 root root 4.0K May  7 03:10 mp1_smu_cmd
-r--------  1 root root 4.0K May  7 03:12 pm_table
-r--------  1 root root 4.0K May  7 03:12 pm_table_size
-r--------  1 root root 4.0K May  7 03:12 pm_table_version
-rw-------  1 root root 4.0K May  7 03:10 rsmu_cmd
-rw-------  1 root root 4.0K May  7 03:10 smn
-rw-------  1 root root 4.0K May  7 03:10 smu_args
-r--------  1 root root 4.0K May  7 03:01 version

# cat /sys/kernel/ryzen_smu_drv/version          
SMU v46.54.0

# cat /sys/kernel/ryzen_smu_drv/mp1_if_version
2

# cat /sys/kernel/ryzen_smu_drv/codename
4

# cat /sys/kernel/ryzen_smu_drv/drv_version
0.0.1

```

Following which, you can run the [test.py script](scripts/test.py) to verify that SMU and SMN
functionality is working:

```
# python3 scripts/test.py
Retrieved SMU Version: v46.54.0
Processor Code Name: Matisse
PM Table: [Supported/Unsupported]
SMN Offset[0x50200]: 0x00001539

Everything seems to be working properly!

```

## Explaining Sysfs Files

#### `/sys/kernel/ryzen_smu_drv/drv_version`

Lists the string-representation of the driver (and thus interface) version. For userspace
applications, they should make sure this version is within the expected range.

#### `/sys/kernel/ryzen_smu_drv/version`

Lists the current SMU firmware version in relation to the currently installed
[AGESA](https://en.wikipedia.org/wiki/AGESA).

The following are several lists of SMU to AGESA versions:

<img src="pics/smu_versions.jpg" width="1000" height="825">

Note: This file returns a string encoded version represented by the "SMU Version" above.

#### `/sys/kernel/ryzen_smu_drv/mp1_if_version`

Lists the interface version for the MP1 mailbox.

This can range from v9 to v13 and is indicated by the following table:

| Value | Interface Version |
|:-----:|:-----------------:|
| 0     | v9                |
| 1     | v10               |
| 2     | v11               |
| 3     | v12               |
| 4     | v13               |
| 5     | Undefined         |

Note: This file returns a string representation of the "Value" field above.

#### `/sys/kernel/ryzen_smu_drv/codename`

Returns a numeric index containing the running processor's codename based on the following
enumeration:

| Hex | Decimal | Code Name      |
|:---:|:-------:|:--------------:|
| 00h | 0       | Unknown        |
| 01h | 1       | Colfax         |
| 02h | 2       | Renoir         |
| 03h | 3       | Picasso        |
| 04h | 4       | Matisse        |
| 05h | 5       | Threadripper   |
| 06h | 6       | Castle Peak    |
| 07h | 7       | Raven Ridge    |
| 08h | 8       | Raven Ridge 2  |
| 09h | 9       | Summit Ridge   |
| 0Ah | 10      | Pinnacle Ridge |
| 0Bh | 11      | Rembrandt      |
| 0Ch | 12      | Vermeer        |
| 0Dh | 13      | Vangogh        |
| 0Eh | 14      | Cezanne        |
| 0Fh | 15      | Milan          |
| 10h | 16      | Dali           |

Note: This file returns 2 characters of the 'Decimal' encoded index.

#### `/sys/kernel/ryzen_smu_drv/rsmu_cmd` or `/sys/kernel/ryzen_smu_drv/mp1_smu_cmd`

This file allows the user to initiate an RSMU or MP1 SMU request. It accepts either an 8-bit or
32-bit command ID that is platform-dependent.

When this file is read, it produces the result on the status of the operation, as a 32 bit
little-endian encoded value.

The SMU returns the following possible values which indicate the result of a command execution:

| Hex | Decimal | Explanation                   |
|:---:|:-------:|:-----------------------------:|
| 00h | 0       | WAITING                       |
| 01h | 1       | OK                            |
| FFh | 255     | FAILED                        |
| FEh | 254     | UNKNOWN COMMAND               |
| FDh | 253     | REJECTED - PREREQUISITE UNMET |
| FCh | 252     | REJECTED - BUSY               |

In addition to this, several custom error codes are also implemented by the driver:

| Hex | Decimal | Explanation                   |
|:---:|:-------:|:-----------------------------:|
| FBh | 251     | COMMAND TIMEOUT               |
| FAh | 250     | INVALID ARGUMENT              |
| F9h | 249     | UNSUPPORTED PLATFORM          |
| F8h | 248     | INSUFFICIENT BUFFER SIZE      |
| F7h | 247     | MEMORY MAP ERROR              |
| F6h | 246     | PCI PROGRAMMING ERROR         |


#### `/sys/kernel/ryzen_smu_drv/smu_args`

When written to, this file accepts 6x 32-bit words (a total of 192 bits) that specify the arguments
used when executing an SMU command.

When read from, it lists either:

- The last values that were written to it before an SMU request was initiated
- The responses from the SMU after a request was completed

Note: All values sent to and read from this file must be in 6x 32-bit words encoded in little-endian
order, arguments numbered from 1 to 6.

#### `/sys/kernel/ryzen_smu_drv/smn`

Allows reading and writing 32 bit values from the SMN address space. To perform an operation, write
a value then read the file for the result.

The amount of bytes written indicates the operation performed:

| Bits Written | Operation | Action Taken                                                                 |
|:------------:|:---------:| ---------------------------------------------------------------------------- |
| 32           | Read      | Reads 32 bit address and returns the result                                  |
| 64           | Write     | Writes the second 32-bit value to the address specified by the first 32 bits |

Note: All values sent to and read from the device must are in little-endian binary format.

#### `/sys/kernel/ryzen_smu_drv/pm_table_size`

On supported platforms, this lists the maximum size of the `/sys/kernel/ryzen_smu_drv/pm_table`
file, in bytes.

Note: File is a 64 bit word encoded in little-endian binary order.

#### `/sys/kernel/ryzen_smu_drv/pm_table_version`

On supported platforms, listed in the table below, this indicates the version of the metrics table.

Each version corresponds to a specific table size and layout that differs across processors.

The following table lists the known characteristics per version:

| Hex      | Platform    | Table Size (Hex) |
|:--------:|:-----------:|:----------------:|
| 0x1E0004 | Raven Ridge | 0x6AC            |
| 0x1E0005 | Raven Ridge | 0x6AC            |
| 0x1E0101 | Raven Ridge | 0x6AC            |
|          |             |                  |
| 0x240802 | Matisse     | 0x7E0            |
| 0x240803 | Matisse     | 0x7E4            |
| 0x240902 | Matisse     | 0x514            |
| 0x240903 | Matisse     | 0x518            |
|          |             |                  |
| 0x2D0803 | Vermeer     | 0x894            |
| 0x380804 | Vermeer     | 0x8A4            |
| 0x380805 | Vermeer     | 0x8F0            |
| 0x2D0903 | Vermeer     | 0x594            |
| 0x380904 | Vermeer     | 0x5A4            |
| 0x380905 | Vermeer     | 0x5D0            |
|          |             |                  |
| 0x370000 | Renoir      | 0x794            |
| 0x370001 | Renoir      | 0x884            |
| 0x370002 | Renoir      | 0x88C            |
| 0x370004 | Renoir      | 0x8AC            |
| 0x370005 | Renoir      | 0x8C8            |
|          |             |                  |
| 0x400005 | Cezanne     | 0x944            |
|          |             |                  |
| 0x2D0008 | Milan       | 0x1AB0           |

Note: File is a 32 bit word encoded in little-endian binary order.

#### `/sys/kernel/ryzen_smu_drv/pm_table`

On supported platforms, this file contains the PM table for the processor, as updated by the SMU.

Note: This file is encoded directly by the SMU and contains an array of 32-bit floating point values
whose structure is determined by the version of the table.

## Module Parameters

The driver supports the following module parameter(s):

#### `smu_timeout_attempts`

When executing an SMU command, either by reading `pm_table` or manually, via `smu_args` and
`smu_cmd`, the driver will retry this many times before considering the command to have timed out.

For example, on slower or busy systems, the SMU may be tied up resulting in commands taking longer
to execute than normal. Allowed range is from `500` to `32768`, defaulting to `8192`.

## Userspace Library

Included in this project is a userspace library, located at [/lib](lib) to allow easy interaction
with the driver. More details of the methods provided may be found in [libsmu.h](lib/libsmu.h).

Example usage of the library is detailed below:

```cpp
#include <stdlib.h>
#include <unistd.h>

#include <libsmu.h>

#define TEST_SMN_ADDR 0x50200

int main(int argc, char** argv) {
    smu_obj_t obj;
    unsigned int result;

    // Userspace library requires root permissions to access driver.
    if (getuid() != 0 && geteuid() != 0) {
        fprintf(stderr, "Program must be run as root.\n");
        exit(-1);
    }

    // Initialize the library for use with the program.
    if (smu_init(&obj) != SMU_Return_OK) {
        fprintf(stderr, "Error initializing userspace library.\n");
        exit(-2);
    }

    printf("Processor Codename: %s\n", smu_codename_to_str(&obj));
    printf("SMU FW: %s\n", smu_get_fw_version(&obj));

    if (smu_read_smn_addr(&obj, TEST_SMN_ADDR, &result) != SMU_Return_OK) {
        fprintf(stderr, "Error reading SMN address: 0x%08x\n", TEST_SMN_ADDR);
        exit(-3);
    }
    printf("SMN [0x%08x]: 0x%08x\n", TEST_SMN_ADDR, result);

    // Cleanup after library use has ended.
    smu_free(&obj);

    return 0;
}
```

N.B. This header file must be compatible with the version of the driver installed.


## Example Usage

For Matisse & Vermeer processors, there are several commands that are known to work. See
[rsmu_commands.md](docs/rsmu_commands.md) for more information.

Commands may be executed either programatically using the library or under a shell by writing to the
designated files.

Example usage via shell:

```sh
# Note: Does not persist across reboots

# Set the running PPT to 142 W (argument in milliwatts)
printf '%0*x' 48 142000 | fold -w 2 | tac | tr -d '\n' | xxd -r -p | sudo tee /sys/kernel/ryzen_smu_drv/smu_args && printf '\x53' | sudo tee /sys/kernel/ryzen_smu_drv/rsmu_cmd

# Set the running TDC to 90 A (argument in milliamps)
printf '%0*x' 48 90000 | fold -w 2 | tac | tr -d '\n' | xxd -r -p | sudo tee /sys/kernel/ryzen_smu_drv/smu_args && printf '\x54' | sudo tee /sys/kernel/ryzen_smu_drv/rsmu_cmd

# Set the running EDC to 140 A (argument in milliamps)
printf '%0*x' 48 140000 | fold -w 2 | tac | tr -d '\n' | xxd -r -p | sudo tee /sys/kernel/ryzen_smu_drv/smu_args && printf '\x55' | sudo tee /sys/kernel/ryzen_smu_drv/rsmu_cmd

# Set the PBO Scalar to 2x
# Calculation: 100 * scalar
printf '%0*x' 48 200 | fold -w 2 | tac | tr -d '\n' | xxd -r -p | sudo tee /sys/kernel/ryzen_smu_drv/smu_args && printf '\x58' | sudo tee /sys/kernel/ryzen_smu_drv/rsmu_cmd
```

As to the rest of the commands, we leave that as an exercise up to the user. :)


debian/changelog
ryzen-smu (0.1.2-1) unstable; urgency=medium

  New revision

 -- Johan Källström <joka90@gmail.com>  Sat, 27 Aug 2022 14:02:53 +0200

ryzen-smu (0.1.1-1) unstable; urgency=medium

  initial debian package release

 -- Johan Källström <joka90@gmail.com>  Sat, 27 Mar 2022 08:18:45 +0100


lib/libsmu.c
/**
 * Ryzen SMU Userspace Library
 * Copyright (C) 2020 Leonardo Gates <leogatesx9r@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **/

#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>

#include "libsmu.h"

#define DRIVER_CLASS_PATH               "/sys/kernel/ryzen_smu_drv/"

#define DRIVER_VERSION_PATH             DRIVER_CLASS_PATH "drv_version"
#define VERSION_PATH                    DRIVER_CLASS_PATH "version"
#define IF_VERSION_PATH                 DRIVER_CLASS_PATH "mp1_if_version"
#define CODENAME_PATH                   DRIVER_CLASS_PATH "codename"

#define SMN_PATH                        DRIVER_CLASS_PATH "smn"
#define SMU_ARG_PATH                    DRIVER_CLASS_PATH "smu_args"
#define RSMU_CMD_PATH                   DRIVER_CLASS_PATH "rsmu_cmd"
#define MP1_SMU_CMD_PATH                DRIVER_CLASS_PATH "mp1_smu_cmd"

#define PM_VERSION_PATH                 DRIVER_CLASS_PATH "pm_table_version"
#define PM_SIZE_PATH                    DRIVER_CLASS_PATH "pm_table_size"
#define PM_PATH                         DRIVER_CLASS_PATH "pm_table"

/* Maximum driver version length defined as "255.255.255\n" */
#define LIBSMU_MAX_DRIVER_VERSION_LEN   12

/* Maximum is defined as: "255.255.255.255\n" */
#define LIBSMU_MAX_SMU_VERSION_LEN      16

static int try_open_path(const char* pathname, int mode, int* fd) {
    int ret = 1;

    *fd = open(pathname, mode);

    // Reset fd to zero to avoid attempting to close a -1 file descriptor.
    if (*fd == -1)
        ret = *fd = 0;

    return ret;
}

static smu_return_val smu_init_parse(smu_obj_t* obj) {
    int ver_maj, ver_min, ver_rev, ver_alt, len, i, c;
    char rd_buf[1024];
    int tmp_fd, ret;

    memset(rd_buf, 0, sizeof(rd_buf));

    // Verify the driver version is expected.
    if (!try_open_path(DRIVER_VERSION_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_DriverNotPresent;

    ret = read(tmp_fd, rd_buf, LIBSMU_MAX_DRIVER_VERSION_LEN);
    close(tmp_fd);

    if (ret < 0)
        return SMU_Return_RWError;

    // The driver version must match the expected exactly.
    if (strcmp(rd_buf, LIBSMU_SUPPORTED_DRIVER_VERSION "\n"))
        return SMU_Return_DriverVersion;

    sscanf(rd_buf, "%d.%d.%d\n", &ver_maj, &ver_min, &ver_rev);
    obj->driver_version = ver_maj << 16 | ver_min << 8 | ver_rev;

    // The version of the SMU **MUST** be present.
    if (!try_open_path(VERSION_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_DriverNotPresent;

    ret = read(tmp_fd, rd_buf, LIBSMU_MAX_SMU_VERSION_LEN);
    close(tmp_fd);

    if (ret < 0)
        return SMU_Return_RWError;

    len = strlen(rd_buf);
    for (i = 0, c = 0; i < len; i++)
        if (rd_buf[i] == '.')
            c++;

    // Depending on the processor, there can be either a 3 or 4 part version segmentation.
    // We account for both.
    switch (c) {
        case 2:
            ret = sscanf(rd_buf, "%d.%d.%d\n", &ver_maj, &ver_min, &ver_rev);
            obj->smu_version = ver_maj << 16 | ver_min << 8 | ver_rev;
            break;
        case 3:
            ret = sscanf(rd_buf, "%d.%d.%d.%d\n", &ver_maj, &ver_min, &ver_rev, &ver_alt);
            obj->smu_version = ver_maj << 24 | ver_min << 16 | ver_rev << 8 | ver_alt;
            break;
        default:
            return SMU_Return_RWError;
    }

    if (ret == EOF || ret < 3)
        return SMU_Return_RWError;

    // Codename must also be present.
    if (!try_open_path(CODENAME_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_DriverNotPresent;

    ret = read(tmp_fd, rd_buf, 3);
    close(tmp_fd);

    if (ret < 0)
        return SMU_Return_RWError;

    obj->codename = atoi(rd_buf);

    if (obj->codename <= CODENAME_UNDEFINED ||
        obj->codename >= CODENAME_COUNT)
        return SMU_Return_Unsupported;

    // MP1 version must also be present.
    if (!try_open_path(IF_VERSION_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_DriverNotPresent;

    // This only specifies an enumeration for the IF version.
    ret = read(tmp_fd, rd_buf, sizeof(rd_buf));
    close(tmp_fd);

    if (ret < 0)
        return SMU_Return_RWError;

    ret = sscanf(rd_buf, "%d\n", (int*)&obj->smu_if_version);
    if (ret == EOF || ret > 3)
        return SMU_Return_RWError;

    // This file doesn't need to exist if PM Tables aren't supported.
    if (!try_open_path(PM_VERSION_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_OK;
    
    ret = read(tmp_fd, &obj->pm_table_version, sizeof(obj->pm_table_version));
    close(tmp_fd);

    if (ret <= 0)
        return SMU_Return_RWError;

    // If the PM table contains a version, a size file MUST exist.
    if (!try_open_path(PM_SIZE_PATH, O_RDONLY, &tmp_fd))
        return SMU_Return_RWError;
    
    ret = read(tmp_fd, &obj->pm_table_size, sizeof(obj->pm_table_size));
    close(tmp_fd);

    if (ret <= 0)
        return SMU_Return_RWError;

    return SMU_Return_OK;
}

smu_return_val smu_init(smu_obj_t* obj) {
    int i, ret;

    memset(obj, 0, sizeof(*obj));

    // Parse constants: SMU Version, Processor Codename, PM Table Size/Version
    ret = smu_init_parse(obj);
    if (ret != SMU_Return_OK)
        return ret;

    // The driver must provide access to these files.
    if (!try_open_path(SMN_PATH, O_RDWR, &obj->fd_smn) ||
        !try_open_path(MP1_SMU_CMD_PATH, O_RDWR, &obj->fd_mp1_smu_cmd) ||
        !try_open_path(SMU_ARG_PATH, O_RDWR, &obj->fd_smu_args))
        return SMU_Return_RWError;

    // RSMU is optionally supported for some codenames.
    if (try_open_path(RSMU_CMD_PATH, O_RDWR, &obj->fd_rsmu_cmd)) {
        // This file may optionally exist only if PM tables are supported AND RSMU as well.
        if (smu_pm_tables_supported(obj) &&
            !try_open_path(PM_PATH, O_RDONLY, &obj->fd_pm_table))
            return SMU_Return_RWError;
    }

    for (i = 0; i < SMU_MUTEX_COUNT; i++)
        pthread_mutex_init(&obj->lock[i], NULL);

    obj->init = 1;

    return SMU_Return_OK;
}

void smu_free(smu_obj_t* obj) {
    int i;

    if (!obj->init)
        return;

    if (obj->fd_smn)
        close(obj->fd_smn);

    if (obj->fd_rsmu_cmd)
        close(obj->fd_rsmu_cmd);

    if (obj->fd_mp1_smu_cmd)
        close(obj->fd_mp1_smu_cmd);

    if (obj->fd_smu_args)
        close(obj->fd_smu_args);

    if (obj->fd_pm_table)
        close(obj->fd_pm_table);

    for (i = 0; i < SMU_MUTEX_COUNT; i++)
        pthread_mutex_destroy(&obj->lock[i]);

    memset(obj, 0, sizeof(*obj));
}

const char* smu_get_fw_version(smu_obj_t* obj) {
    static char fw[32] = { 0 };

    if (!obj->init)
        return "Uninitialized";

    // Determine if this is a 24-bit or 32-bit version and show it accordingly.
    if (obj->smu_version & 0xff000000) {
        sprintf(fw, "%d.%d.%d.%d",
            (obj->smu_version >> 24) & 0xff, (obj->smu_version >> 16) & 0xff,
            (obj->smu_version >> 8) & 0xff, obj->smu_version & 0xff);
    }
    else
        sprintf(fw, "%d.%d.%d",
            (obj->smu_version >> 16) & 0xff, (obj->smu_version >> 8) & 0xff,
            obj->smu_version & 0xff);

    return fw;
}

smu_return_val smu_read_smn_addr(smu_obj_t* obj, unsigned int address, unsigned int* result) {
    unsigned int ret;

    // Don't attempt to execute without initialization.
    if (!obj->init)
        return SMU_Return_Failed;

    pthread_mutex_lock(&obj->lock[SMU_MUTEX_SMN]);

    lseek(obj->fd_smn, 0, SEEK_SET);
    ret = write(obj->fd_smn, &address, sizeof(address));

    if (ret != sizeof(address))
        goto BREAK_OUT;

    lseek(obj->fd_smn, 0, SEEK_SET);
    ret = read(obj->fd_smn, result, sizeof(*result));

BREAK_OUT:
    pthread_mutex_unlock(&obj->lock[SMU_MUTEX_SMN]);

    return ret == sizeof(unsigned int) ? SMU_Return_OK : SMU_Return_RWError;
}

smu_return_val smu_write_smn_addr(smu_obj_t* obj, unsigned int address, unsigned int value) {
    unsigned int buffer[2], ret;

    // Don't attempt to execute without initialization.
    if (!obj->init)
        return SMU_Return_Failed;

    // buffer[0] contains the destination write target.
    // buffer[1] contains the value to write to the address.
    buffer[0] = address;
    buffer[1] = value;

    pthread_mutex_lock(&obj->lock[SMU_MUTEX_SMN]);

    lseek(obj->fd_smn, 0, SEEK_SET);
    ret = write(obj->fd_smn, buffer, sizeof(buffer));

    pthread_mutex_unlock(&obj->lock[SMU_MUTEX_SMN]);

    return ret == sizeof(buffer) ? SMU_Return_OK : SMU_Return_RWError;
}

smu_return_val smu_send_command(smu_obj_t* obj, unsigned int op, smu_arg_t* args,
    enum smu_mailbox mailbox) {
    unsigned int ret, status, fd_smu_cmd;

    // Don't attempt to execute without initialization.
    if (!obj->init)
        return SMU_Return_Failed;

    switch (mailbox) {
        case TYPE_RSMU:
            fd_smu_cmd = obj->fd_rsmu_cmd;
            break;
        case TYPE_MP1:
            fd_smu_cmd = obj->fd_mp1_smu_cmd;
            break;
        default:
            return SMU_Return_Unsupported;
    }

    // Check if fd is valid.
    if (!fd_smu_cmd)
        return SMU_Return_Unsupported;

    pthread_mutex_lock(&obj->lock[SMU_MUTEX_CMD]);

    lseek(obj->fd_smu_args, 0, SEEK_SET);
    ret = write(obj->fd_smu_args, args->args, sizeof(*args));

    if (ret != sizeof(*args)) {
        ret = SMU_Return_RWError;
        goto BREAK_OUT;
    }

    lseek(fd_smu_cmd, 0, SEEK_SET);
    ret = write(fd_smu_cmd, &op, sizeof(op));

    if (ret != sizeof(op)) {
        ret = SMU_Return_RWError;
        goto BREAK_OUT;
    }

    // Commands should be completed instantly as the driver attempts to continuously
    //  execute it till a timeout has occurred and immediately updates the result.
    // Therefore it shouldn't be necessary to apply any sort of waiting here.
    lseek(fd_smu_cmd, 0, SEEK_SET);
    ret = read(fd_smu_cmd, &status, sizeof(status));

    if (ret != sizeof(status))
        ret = SMU_Return_RWError;
    else
        ret = status;

    if (ret == SMU_Return_OK) {
        lseek(obj->fd_smu_args, 0, SEEK_SET);
        ret = read(obj->fd_smu_args, args->args, sizeof(args->args)) == sizeof(args->args)
            ? SMU_Return_OK
            : SMU_Return_RWError;
    }

BREAK_OUT:
    pthread_mutex_unlock(&obj->lock[SMU_MUTEX_CMD]);

    return ret;
}

smu_return_val smu_read_pm_table(smu_obj_t* obj, unsigned char* dst, size_t dst_len) {
    int ret;

    // Don't attempt to execute without initialization.
    if (!obj->init)
        return SMU_Return_Failed;

    if (dst_len != obj->pm_table_size)
        return SMU_Return_InsufficientSize;

    pthread_mutex_lock(&obj->lock[SMU_MUTEX_PM]);

    lseek(obj->fd_pm_table, 0, SEEK_SET);
    ret = read(obj->fd_pm_table, dst, obj->pm_table_size);

    if (ret != obj->pm_table_size)
        ret = SMU_Return_RWError;
    else
        ret = SMU_Return_OK;

    pthread_mutex_unlock(&obj->lock[SMU_MUTEX_PM]);

    return ret;
}

const char* smu_return_to_str(smu_return_val val) {
    switch (val) {
        case SMU_Return_OK:
            return "OK";
        case SMU_Return_Failed:
            return "Failed";
        case SMU_Return_UnknownCmd:
            return "Unknown Command";
        case SMU_Return_CmdRejectedPrereq:
            return "Command Rejected - Prerequisite Unmet";
        case SMU_Return_CmdRejectedBusy:
            return "Command Rejected - Busy";
        case SMU_Return_CommandTimeout:
            return "Command Timed Out";
        case SMU_Return_InvalidArgument:
            return "Invalid Argument Specified";
        case SMU_Return_Unsupported:
            return "Unsupported Platform Or Feature";
        case SMU_Return_InsufficientSize:
            return "Insufficient Buffer Size Provided";
        case SMU_Return_MappedError:
            return "Memory Mapping I/O Error";
        case SMU_Return_PCIFailed:
            return "PCIe Programming Error";
        case SMU_Return_DriverNotPresent:
            return "SMU Driver Not Present Or Fault";
        case SMU_Return_RWError:
            return "Read Or Write Error";
        case SMU_Return_DriverVersion:
            return "SMU Driver Version Incompatible With Library Version";
        default:
            return "Unspecified Error";
    }
}

const char* smu_codename_to_str(smu_obj_t* obj) {
    switch (obj->codename) {
        case CODENAME_CASTLEPEAK:
            return "CastlePeak";
        case CODENAME_COLFAX:
            return "Colfax";
        case CODENAME_MATISSE:
            return "Matisse";
        case CODENAME_PICASSO:
            return "Picasso";
        case CODENAME_PINNACLERIDGE:
            return "Pinnacle Ridge";
        case CODENAME_RAVENRIDGE2:
            return "Raven Ridge 2";
        case CODENAME_RAVENRIDGE:
            return "Raven Ridge";
        case CODENAME_RENOIR:
            return "Renoir";
        case CODENAME_SUMMITRIDGE:
            return "Summit Ridge";
        case CODENAME_THREADRIPPER:
            return "Thread Ripper";
        case CODENAME_REMBRANDT:
            return "Rembrandt";
        case CODENAME_VERMEER:
            return "Vermeer";
        case CODENAME_VANGOGH:
            return "Van Gogh";
        case CODENAME_CEZANNE:
            return "Cezanne";
        case CODENAME_MILAN:
            return "Milan";
        case CODENAME_DALI:
            return "Dali";
        case CODENAME_LUCIENNE:
            return "Lucienne";
        default:
            return "Undefined";
    }
}

unsigned int smu_pm_tables_supported(smu_obj_t* obj) {
    return obj->pm_table_size && obj->pm_table_version;
}
